src/main/java/software/amazon/event/kafkaconnector/offloading/S3EventBridgeEventDetailValueOffloading.java

```diff
@@ -0,0 +1,224 @@
+            .collect(partitioningBy(EventBridgeResult::isSuccess));
+
+    var success = result.get(true).stream().map(EventBridgeResult::success).collect(toList());
+    var errors = result.get(false).stream().map(EventBridgeResult::failure).collect(toList());
+
+    return new EventBridgeMappingResult(success, errors);
+  }
+
+  private EventBridgeResult<PutEventsRequestEntry> apply(
+      final MappedSinkRecord<PutEventsRequestEntry> item) {
+
+    var sinkRecord = item.getSinkRecord();
+    var putEventsRequestEntry = item.getValue();
+
+    try {
+      var documentContext = JsonPath.parse(putEventsRequestEntry.detail(), jsonPathConfiguration);
+
+      var value = documentContext.read(jsonPathRemove);
+      if (value != null) {
+        var payload =
+            value instanceof Map || value instanceof List
+                ? JsonPath.parse(value).jsonString()
+                : value.toString();
+        documentContext
+            .delete(jsonPathRemove)
+            .put(jsonPathAdd, "dataref", s3ArnOf(putS3Object(payload)));
+      }
+      documentContext.put(jsonPathAdd, "datarefJsonPath", jsonPathExp);
+
+      return success(
+          sinkRecord, putEventsRequestEntry.copy(it -> it.detail(documentContext.jsonString())));
+
+    } catch (final ThrowingFunctionApplyException e) {
+      var unwrapped = e.getCause();
+      if (unwrapped instanceof ExecutionException
+          || unwrapped instanceof InterruptedException
+          || unwrapped instanceof TimeoutException) {
+
+        var cause = unwrapped.getCause();
+        if (cause instanceof S3Exception && ((S3Exception) cause).statusCode() < 500) {
+          return failure(sinkRecord, panic(e));
+        }
+        return failure(sinkRecord, retry(e));
+      }
+      return failure(sinkRecord, panic(unwrapped));
+    }
+  }
+
+  private String s3ArnOf(UUID uuid) {
+    return format("arn:aws:s3:::%s/%s", bucketName, uuid);
+  }
+
+  private UUID putS3Object(final String payload) {
+    var durationRef = new AtomicReference<Duration>();
+    var s3ObjectId =
+        s3ObjectKeyCache.computeIfAbsent(
+            sha512Of(payload),
+            ThrowingFunction.wrap(
+                (sha512) -> {
+                  var id = idGenerator.get();
+                  var request =
+                      PutObjectRequest.builder().bucket(bucketName).key(id.toString()).build();
+
+                  var body = fromString(payload, UTF_8);
+
+                  var start = OffsetDateTime.now(UTC);
+                  client.putObject(request, body).get(SDK_TIMEOUT, MILLISECONDS);
+                  durationRef.set(Duration.between(start, OffsetDateTime.now(UTC)));
+
+                  return id;
+                }));
+
+    var duration = durationRef.get();
+    logger.trace(
+        "found JSON at '{}' cache={} durationMillis={}",
+        jsonPathExp,
+        duration == null ? "HIT" : "MISS",
+        duration == null ? "-" : duration.toMillis());
```

Comment:
question: this is what I see in the log now

```console
connect-1  | [2024-05-07 06:56:55,307] TRACE [eventbridge-s3-test|task-0] [@ec77526-dirty] found JSON at '$.detail.value.image' cache=HIT durationMillis=- (software.amazon.event.kafkaconnector.offloading.S3EventBridgeEventDetailValueOffloading:216)
```

The `-` looks a bit weird for a duration. Was wondering if default should be `0` then? WDYT?

Reply:
Will change this with your suggestion.

Reply:
Every cache 'HIT' has now an log line like:
```
found JSON at '$.detail.value' cache=HIT durationMillis=0
``